### PR TITLE
feat: Support GCM encryption for Gree devices

### PR DESCRIPTION
**Pull request by cmroche**
Adds support for GCM encryption used on some (newer) Gree devices. Since it's not clear yet how to identify these devices by their announce information, the older AES ECB key will be tried first, then attempt the AES GCM key.

Changes:
- [x] Finish Cipher AES-GCM support
- [x] Reduce 120 timeout on binding wait only
- [x] Add tests
- [x] Verify in Home Assistant

### CODE REVIEW

File: greeclimate/device.py [hunk 1]
```diff
@@ -207,28 +216,44 @@ async def bind(self, key=None):
...
+        if not self.device_cipher:
             raise DeviceNotBoundError
         else:
-            self._logger.info("Bound to device using key %s", self.device_key)
+            self._logger.info("Bound to device using key %s", self.device_cipher.key)
+
+    async def __bind_internal(self, cipher_type: type[Union[CipherV1, CipherV2]]):
+        """Internal binding procedure, do not call directly"""
+        default_key = GENERIC_CIPHERS_KEYS.get(cipher_type)
+        await self.send(self.create_bind_message(self.device_info), cipher=cipher_type(default_key))
```
Comment (engrbm87) [inline greeclimate/device.py:243]: May I suggest that you define the generic key as a class attribute inside the CipherV1 and CipherV2 classes. In the `__init__` you can define the key as optional and use the generic key if `key` is None.

File: greeclimate/network.py [hunk 2]
```diff
@@ -45,18 +38,19 @@ class IPInterface:
 class DeviceProtocolBase2(asyncio.DatagramProtocol):
     """Event driven device protocol class."""
 
+    _transport: Union[asyncio.transports.DatagramTransport, None] = None
+    _cipher: Union[CipherBase, None] = None
```
Comment (engrbm87) [inline greeclimate/network.py:42]: I did a quick test using this branch. I can confirm that the broadcast message is encrypted with V1 Cipher. This means initially the cipher should be V1. During binding, if it times out, we should try V2 Cipher.

File: greeclimate/device.py [hunk 3]
```diff
@@ -155,6 +161,13 @@ class Device(DeviceProtocol2, Taskable):
...
     """
 
+    """ Device properties """
+    hid = None
+    version = None
+    check_version = True
+    _properties = {}
+    _dirty = []
+
     def __init__(self, device_info: DeviceInfo, timeout: int = 120, loop: AbstractEventLoop = None):
```
Comment (engrbm87) [inline greeclimate/device.py:171]: Setting timeout to 120 means the binding attempt has to wait 120 seconds before it times out and tries the new cipher.
I am not if we should lower the timeout here or if it is possible to set a different timeout for the binding itself.

Comment (cmroche) [reply greeclimate/device.py:171]: This is the socket timeout, it probably should stay at this value, but I can add another timeout to the bind response specifically as an exception.

File: greeclimate/network.py (hunk 2 above)
Comment (cmroche) [reply greeclimate/network.py]: That was a lucky guess on my side. Thanks for checking.

**Comment by cmroche** [comment]
@engrbm87 Are you comfortable to test this branch with your HA to confirm? While I have a bit more work to do for tests to pass, I think this should be functional now.

**Comment by engrbm87** [comment]
> @engrbm87 Are you comfortable to test this branch with your HA to confirm? While I have a bit more work to do for tests to pass, I think this should be functional now.

Nice work, I noticed one thing still missing. The initial binding timeout. It is still waiting for 120 seconds.

**Comment by angelitoo776** [comment]
Hi!

I was doing some work to include GCM encryption support and I see here it almost done (and much better implemented).
There is something that I can help on this?

**Comment by cmroche** [comment]
> Hi!
> 
> I was doing some work to include GCM encryption support and I see here it almost done (and much better implemented). There is something that I can help on this?

Testing this PR if you are up for it please. Otherwise I will aim to finish up and merge this week.

**Comment by cmroche** [comment]
> Nice work, I noticed one thing still missing. The initial binding timeout. It is still waiting for 120 seconds.

Yup, wanted to get the cipher stuff out of the way, I'll look at this ridiculous timeout in the next day or two.

**Comment by cmroche** [comment]
@engrbm87 @angelitoo776 I'm getting ready to release this into main, perhaps you could do some quick testing and confirm. The bind timeout for V2 has been adjusted down to 10 seconds now too.

**Comment by engrbm87** [comment]
Hello @cmroche , thanks for the updated bind timeout. One thing I noticed during testing is that the `device.update_state()` is not awaited correctly. Meaning that after the update request packet is sent we are not waiting for the response before executing other functions.
To clarify what is happening in HA. The coordinator is requesting a refresh which triggers a device update. The coordinator is not waiting for the response to be processed and considers that the coordinator data update is completed. When the climate entity is initialized without the response date being processed, the temperature unit conversion is not calculated correctly (getting negative temperature).
I am not sure where exactly is the library waiting for the response before finishing the `update_state()` function and returning.

**Comment by cmroche** [comment]
> sure where exactly is the library waiting for the response before finishing the `update_state()` function and returning.

This is by design. The goal is to stop the lock-step approach to device comms over the socket. Since networking is inherently unreliable this should lead to a more robust overall design.

That said, the issue is one I have not seen likely because of my ideal network conditions. BUT, I think it would be reasonable to mark the device as unavailable until the first state update, or to block in the binding process until there is an initial state.

**Comment by cmroche** [comment]
@engrbm87 Does the negative temperature resolve itself once an update is received, or is it persistent?

**Comment by engrbm87** [comment]
> @engrbm87 Does the negative temperature resolve itself once an update is received, or is it persistent? I believe it should resolve once the update comes in, since the correction is computed on the incoming state update from the device.

I just noticed PR# 121041. I updated my test component based on that and now the temperatures are showing correct. I still think the coordinator should wait until the response is processed during initial setup.

**Pull request merged**